### DISCUSSION TITLE
svg file support in OpenPDF

### DISCUSSION THREAD

**Discussion by pronabb** [LibrePDF/OpenPDF#1078] [Q&A]
Hi,
We want to use SVG files to add icons in PDF. Is there any standard way to do that using OpenPDF?

**Comment by peda** [comment]
In the old iText version you could use PDF files in the src Attribute of img Tags - if done so it does include the vector graphics inside this other PDF - that's what I've used in the past to workaround adding SVGs

**Comment by V-F** [comment]
You can try to use [Apache Batik](https://mvnrepository.com/artifact/org.apache.xmlgraphics) library for that. It’s need to make the following steps:
1.	Create `org.w3c.dom.svg.SVGDocument:`
```
SAXSVGDocumentFactory factory = new org.apache.batik.anim.dom.SAXSVGDocumentFactory(XMLResourceDescriptor.getXMLParserClassName());
factory.createSVGDocument("", InputStream);
```
2.	[Create ](https://stackoverflow.com/questions/44177676/how-to-get-svg-node-at-coordinates-with-batik) a `org.apache.batik.gvt.GraphicsNode` from the `org.w3c.dom.svg.SVGDocument`:
```
org.apache.batik.bridge.UserAgent userAgent = new 
org.apache.batik.bridge.UserAgentAdapter();
org.apache.batik.bridge.DocumentLoader loader = new org.apache.batik.bridge.DocumentLoader(userAgent);
// Notice, that you should use org.apache.batik.bridge.svg12.SVG12BridgeContext.SVG12BridgeContext for the svg version 1.2

org.apache.batik.bridge.BridgeContext context = new org.apache.batik.bridge.BridgeContext(userAgent, loader);
context.setDynamicState(
org.apache.batik.bridge.BridgeContext.DYNAMIC);

org.apache.batik.bridge.GVTBuilder builder = new 
org.apache.batik.bridge.GVTBuilder();
org.apache.batik.gvt.GraphicsNode rootGraphicsNode = builder.build(context, SVGDocument);
```
3.	Define the size attributes in the SVGDocument if they are absent, sinse GVTBuilder analyzes the size attributes of the SVGDocument later by painting (otherwise exception will be thrown).
4.	Create PDFTemplate:
`PdfTemplate template = com.lowagie.text.pdf.PdfTemplate.createTemplate(PdfWriter writer, float svgImageWidthInPx, float svgImageHeightInPx);`
5.	Create graphics from this template and draw the graphic node with it:
```
Graphics2D g2d = template.createGraphics(template.getWidth(), template.getHeight(), FontMapper);
try {
    rootGraphicsNode.paint(g2d);
} finally {
    g2d.dispose();
}
```
6.	Create  `com.lowagie.text.Image` object that you can use where you want.
`new com.lowagie.text.ImgTemplate.ImgTemplate (template);`

Notice, that you can use the same instance of the image object as you need to use the same image in pdf file. It helps to reduce the size of the generated pdf file.

**Comment by RangerMak** [comment]
This solution is partially works for me.
But result image is cut off by height. So i have to increase height for template creation.
Getting of size code:
```
float svgImageWidthInPx = (float) rootGraphicsNode.getPrimitiveBounds().getWidth();
float svgImageHeightInPx = (float) rootGraphicsNode.getPrimitiveBounds().getHeight();
```
Is i made something wrong?

**Comment by softwaremagico** [comment]
@RangerMak I have notice, that `rootGraphicsNode.getPrimitiveBounds()` gets the size of the content from the SVG, not the viewbox. To have the correct size and avoid the image to be cut, I need to add a rectangle as a "background" with the same size that the desired viewbox size. Then, all image is shown on the PDF without cutting it. 

From my tests, you can use a transparent rectangle, and the size will be obtained correctly, without adding any undesired background color to the image.

**Comment by AidMen** [comment]
Are there any news if support for SVGs will be implemented in OpenPDF natively?

**Comment by andreasrosdalw** [comment]
The PDF standard does not natively support SVG (Scalable Vector Graphics) files.

**Comment by krissrex** [comment]
Here's another example, using [jsvg](https://github.com/weisJ/jsvg) instead of batik https://github.com/LibrePDF/OpenPDF/issues/777#issuecomment-3272412145 .


Note that I do not use the Document API, but rather the raw `directContent` and `PdfGraphics2D` to render a full A4 page-sized SVG document.

```svg
<svg
   width="595"
   height="842"
   viewBox="0 0 595 842"
...
```
